<center> <h2> DS 3000 - Spring 2019</h2> </center>
<center> <h3> DS Report </h3> </center>


<center> <h3>Predicting Oil Exposure in Shore Birds</h3> </center>
<center><h4>Tom Cairns, Ben Kosiborod, Gabe Deacon, Amy Tang</h4></center>


<hr style="height:2px; border:none; color:black; background-color:black;">

#### Executive Summary:
Oil spills affect the health of shore birds, yet it may be difficult for researchers to determine the proper physiological markers to look for when determining whether a given shorebird was affected. We attempted to solve this knowledge gap by constructing a model to help predict whether a shorebird was impacted by an oil spill. To do this, we took data from a Deepwater Horizon Oil Spill study that measured various physiological markers in captured shorebirds that were marked as either "Reference" birds (not affected by an oil spill), or "Impacted". We fed this data to a supervised machine learning classification algorithm and tuned the algorithm to be able to predict with a reasonable amount of accuracy whether a given shorebird was affected by an oil spill. We cleaned up and explored our data, which we then used to construct our model and evaluate its performance with some of our data, and tested the model with the rest of the data (i.e. any data we didn't use to train the model). Our hypotheses turned out to be somewhat supported by the data, but we discovered that this varies depending on which species is the focus. The most accurate ML classification algorithm turned out to be a Decision Tree algorithm, which is able to predict with roughly `92%` accuracy whether a given shore bird was affected by an oil spill.

<hr style="height:2px; border:none; color:black; background-color:black;">

## Outline
1. <a href='#1'>INTRODUCTION</a>
2. <a href='#2'>METHOD</a>
3. <a href='#3'>RESULTS</a>
4. <a href='#4'>DISCUSSION</a>

<a id="1"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 1. INTRODUCTION

###Problem Statement:

We are using supervised machine learning classification algorithms to look at physiological markers in different species of shore birds and identify the markers most impacted by oil spills. We are seeing if from these markers we can predict if a shore bird were exposed to oil. We want to see which markers are the best to predict whether a bird has been exposed to oil. 


###Significance of the Problem:

The DeepWater Horizon spill was not the first major oil spill, nor will it be the last. Oil spills can have far reaching effects on ecosystems. This project is important because it can provide insights into how shore birds are affected by oil spills and the physiological markers. Dramatic changes in physiology can have long term effects, such as affecting fecundity, so the results of this model can help predict how shore birds are impacted by an oil spill. It can also be used to see how far the impacts of an oil spill can be seen. Since many birds migrate to or through the Gulf of Mexico, many more birds may have been impacted by the spill than initially thought. Therefore, finding a simple way to test a bird for oil exposure is imperative to discovering how far the effects of the oil spill traveled. 

The external effects of oil have been vastly researched, showing how oil 
removes the water repellent properties of feathers (Morandin and O’Hara 2016). This has an effect on buoyancy, can lead to behavioral changes, and can cause impaired flight. Exposure and ingestion of oil have been shown to have metabolic effects on marine birds, when not in lethal doses (Morandin and O’Hara 2016). Birds exposed to oil have been shown to be less able to cope with other stressors than birds not exposed to oil (Lattin et al. 2014). Polycyclic aromatic hydrocarbons (PAHs) are a chemical compound that are derived from fossil fuels. These PAHs are toxic at low concentrations and can be transferred up the food chain. Even when this compound is not consumed in lethal doses, it can have serious physiological effects like liver damage, anemia, immune suppression, gastrointestinal damage etc. (Paruk et al. 2016). Seabirds may be more at risk of the harmful effects than terrestrial birds of oil due to their longer time spent maturing and the fact that many species only lay one egg per mating season (Morandin and O’hara 2016). This means the species would be less able to adapt and evolve to these harmful conditions. The effects of oil pollution may also cause lasting effects in the reproductive ability of birds exposed to oil, so the population of shorebirds may be impacted for more than one generation.

  In our project we are looking at physiological markers in different species of shore birds to identify the markers most impacted by oil spills. We are seeing if we can accurately predict if a bird was exposed to oil pollution. A better understanding of physiological markers in shore birds helps researchers to target certain markers to classify whether a bird has been involved in an oil spill; this can help identify unknown oil spills.  

```
Citations:

Lattin, C. R., Ngai, H. M., & Romero, L. M. (2014). Evaluating the Stress
  Response as a Bioindicator of Sub-Lethal Effects of Crude Oil Exposure in
  Wild House Sparrows (Passer domesticus). PLoS ONE, 9(7).

Morandin, L. A., & O’Hara, P. D. (2016). Offshore oil and gas, and operational
  sheen occurrence: is there potential harm to marine birds? Environmental
  Reviews, 24(3), 285–318. https://doi-org.ezproxy.neu.edu/10.1139/er-2015-0086

Paruk, J. D., Adams, E. M., Uher-Koch, H., Kovach, K. A., Long, D. t., Perkins
  C., . . . Evers, D. C. (2016). Polycyclic aromatic hydrocarbons in blood
  related to lower body mass in common loons. Sci Total Environ, 565, 360-368.
  doi:10.1016/j.scitotenv.2016.04.150
```

### Questions:

From the previously mentioned problem and its importance, we set out to answer the following questions: 
* Can the physiological markers tested accurately predict whether a shore bird has been exposed to oil pollution? 
* Is there a decrease in reticulocyte counts in impacted birds, indicating anemia? 
* Is there a change in metabolic markers (uric acid, cholesterol levels, glucose levels, etc) in the impacted shore birds? 
* Will the logistic regression algorithm be useful, or will one of the other classification algorithms we learned be more accurate?
    
### Hypotheses:
* We hypothesize that an increase in PAHs will indicate that a bird was impacted by oil pollution. 
* We hypothesize that there will be a decrease in reticulocyte counts for impacted birds. 
* We hypothesize that there will be a decrease in the metabolic marker, uric acid, in impacted shore birds. 


<a id="2"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 2. METHOD

### 2.1. Data Acquisition

The data is from a report, Deepwater Horizon (MC 252) Oil Spill Natural Resource Damage Assessment and Restoration, and in CSV format. The CSV can be found [here](https://raw.githubusercontent.com/tomcairns1/DS3000FinalProject/master/DWH_NRDA_AvianPhysiology.xls), while a document describing the data can be found [here](https://github.com/tomcairns1/DS3000FinalProject/blob/master/DWH_NRDA_AvianPhysiology_DataReport.pdf). We originally got this data and its description from [here](https://catalog.data.gov/dataset/dwh-nrda-blood-parameters-as-a-measure-of-physiological-injury-to-oiled-birds), but recently we have not been able to access that link. The citation of the study that compiled this data is:
```
Fallon, J. A., et al. (2018). "Hematological indices of injury to lightly oiled birds from the Deepwater Horizon oil spill." Environ Toxicol Chem 37(2): 451-461.
```
A direct link to a PDF version of the study can be found [here](http://www.briloon.org/uploads/Library/item/526/file/Fallon_et_al-2017-Environmental_Toxicology_and_Chemistry.pdf).

The dataset contains detailed capture information including date, location (State, longitude/latitude, capture site) and information about the bird such as its age, weight, and band-number. The dataset also contains various blood measurements and physiological markers describing each bird. The relevant variables in our study are:
- Classification: if the bird was impacted or not by an oil spill
- Physiological markers such as Plasma Haptoglobin (g/dl), Packed Cell Volume, etc. 


### 2.2. Variables
We hypothesized that the birds affected by an oil spill will have different physiological markers (such as PAH (ng/mL), Heinz Body Counts, Reticulocytes, etc.). 

These **physiological markers will be the independent variables**, and the **classification (whether the bird was a reference or impacted) will serve as the dependent variable**. In our model, the features will be the IVs (physiological markers), and the target will be the DV (Classification). 


### 2.3. Data Analysis
This is a supervised learning classification problem, and our predictive model aims to use physiological markers (such as PAH (ng/mL), Heinz Body Counts, Reticulocytes, etc.) to predict if a bird has been impacted by an oil spill (classification). We used classification algorithms because the goal of our analysis is to be able to classify a bird as "impacted" or "not impacted" by an oil spill based on physiological markers. We decided to test _all_ of the classification algorithms we learned over the course of this semester, because it is trivial to train/test them all, and we did not have a strong inclination to use one in particular.  

We tested 5 classifiers: 
* K-Nearest Neighbors-based classification (KNeighborsClassifier) 
- Linear Support Vector Classification (LinearSVC)
- Gaussian Naive Bayes (GaussianNB)
- Decision Tree classifier (DecisionTreeClassifier)
- Logistic Regression classifier (LogisticRegression)

We decided to use the Decision Tree classifier due to its prediction accuracy and low error rate. 


<a id="3"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 3. RESULTS

### 3.1. Data Wrangling
* Perform simple data cleaning (delete extra columns, deal with NA values, etc.)
* Perform data wrangling to extract your features and target values (e.g., grouping your dataframe by columns, applying functions to format dataframes, etc.)
* Preprocess your variables (e.g., scaling/transforming feature variables to normalize them)
* Feature extraction (dummy variables, new features from existing features, etc.)
* Use one feature selection technique to select a subset of your original features


In [0]:
# All import statements we will need
import pandas as pd
import numpy as np
import math
import plotly.express as px
from scipy import stats
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import MinMaxScaler
from sklearn.feature_selection import SelectFromModel
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from sklearn.svm import LinearSVC
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

In [0]:
# Pull the data from the github repository where it's stored
url = 'https://raw.githubusercontent.com/tomcairns1/DS3000FinalProject/master/DWH_NRDA_AvianPhysiology.xls'
df = pd.read_excel(url, error_bad_lines=False)
df

Band Number                 Date  ... H:L ratio Serum Condition
0     1055-04860  2010-10-13 00:00:00  ...   1.91176          Normal
1     1106-09098  2010-10-13 00:00:00  ...   2.52174          Normal
2     1106-14620  2010-09-25 00:00:00  ...   2.63636          Normal
3     1106-14621  2010-09-25 00:00:00  ...   2.90909          Normal
4     1106-14622  2010-09-25 00:00:00  ...   2.38462          Normal
...          ...                  ...  ...       ...             ...
1251  2551-07895  2010-09-24 00:00:00  ...        NR              NR
1252  2551-07896  2010-09-24 00:00:00  ...        NR              NR
1253  2551-07897  2010-09-24 00:00:00  ...        NR              NR
1254  2551-07898  2010-09-24 00:00:00  ...        NR              NR
1255  2551-07899  2010-09-24 00:00:00  ...        NR              NR

[1256 rows x 67 columns]

In [0]:
# drop all rows where all columns from "Heinz Body Count" to the end of the DataFrame
# are None values (includes "NR", "ND", and " ")
columns_to_analyze = list(df.columns[14:])
clean_df = df.replace(" ", np.NaN).replace("NR", np.NaN).replace("ND", np.NaN).dropna(how="all", subset=columns_to_analyze).reset_index().drop("index", axis=1)
clean_df

Band Number                 Date  ... H:L ratio Serum Condition
0    1055-04860  2010-10-13 00:00:00  ...  1.911765          Normal
1    1106-09098  2010-10-13 00:00:00  ...  2.521739          Normal
2    1106-14620  2010-09-25 00:00:00  ...  2.636364          Normal
3    1106-14621  2010-09-25 00:00:00  ...  2.909091          Normal
4    1106-14622  2010-09-25 00:00:00  ...  2.384615          Normal
..          ...                  ...  ...       ...             ...
613  2551-07857  2010-09-21 00:00:00  ...  0.278689    Slides only.
614  2551-07861  2010-09-22 00:00:00  ...  0.231884    Slides only.
615  2551-07864  2010-09-22 00:00:00  ...  0.750000    Slides only.
616  2551-07880  2010-09-23 00:00:00  ...  0.366667     Slides only
617  2551-07881  2010-09-23 00:00:00  ...  1.066667     Slides only

[618 rows x 67 columns]

In [0]:
# dropping the SSSP species as there is no usable data
# starting from the "Heinz Body Count" column, so there
# is not enough data to give to a model such that it would
# be able to reliably predict that a bird in this species
# is a reference or impacted
clean_df = clean_df[clean_df["Species"] != "SSSP"]

In [0]:
# for Heinz Body Count, missing values are replaced with 0
# because background research shows that this value is typically 0
# and most of the other values in this column were 0

clean_df["Heinz Body Count"] = clean_df["Heinz Body Count"].apply(lambda x: 0 if math.isnan(x) else x)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [0]:
# Functions for cleaning up data

# aggregrates the median for a given column by species and classification; 
# replaces the nan values in the given column by the median based on the 
# species and classification
def replaceNanWithMedian(column, df_to_clean): 
  res_df = df_to_clean.copy()
  clean_df_grouped = df_to_clean.groupby(["Species", "Classification"])
  medians = clean_df_grouped[column].median().to_dict()
  for index, row in df_to_clean.iterrows():
    if math.isnan(row[column]):
      res_df.at[index, column] = medians.get((row["Species"], row["Classification"]))
  return res_df

# drops the given column from the dataframe if there is an 
# aggregrated median of nan for the species and classification
def dropUnusableCols(column, df_to_clean):
  res_df = df_to_clean.copy()
  clean_df_grouped = df_to_clean.groupby(["Species", "Classification"])
  medians = clean_df_grouped[column].median().to_dict()
  for key in medians.keys():
    if math.isnan(medians.get(key)):
      res_df = res_df.drop(columns=column)
      break
  return res_df

In [0]:
# Here, we are cleaning continous data in the dataframe:

# categorical columns that we do not want to clean, as right now we are
# only dealing with continuous data
cat_columns = ["Band Number", "Date", "Capture Site", "State", "Latitude", "Longitude", "Site Characteristics", "Classification", 
                     "Species", "Age", "Bird Weight (g)", "Visible Oil Assessment", "UV Light Oil Assessment", "Oil at Capture Site", 
                     "Heinz Body Count", "Buffy Coat", "Polychromasia", 	"Anisocytosis", "RBC Count (million/cubic mm)","Serum Condition"]

# creating list of columns that are continous, based on the categorical
# columns above
columns_median_fix = clean_df.columns.to_list() 
for col in cat_columns:
  columns_median_fix.remove(col)

# replaces the NaN in the columns with continuous data with the median,
# based on function in above cell 
new_ = clean_df.copy()
for feature in columns_median_fix:
  new_ = replaceNanWithMedian(feature, new_)

# removing columns where any aggregrate median is nan
for feature in columns_median_fix:
  new_ = dropUnusableCols(feature, new_)

clean_df = new_

In [0]:
# Cleaning up categorical data in the DataFrame:

# removing Site Characteristics because it is not possible to clean 
# this column since we can only fill NaN with Unknowns 
# (doesn't make sense in the long run)
clean_df.drop("Site Characteristics", axis=1, inplace=True)

# Removing Age and Bird Weight (g)
# Age doesn't appear to be useful for determining if the bird is from an impacted site
# Bird Weight had some clearly incorrect values, and fixing them is not possible
fig = px.scatter(clean_df, x="Age", y="Bird Weight (g)", color="Species")
fig.show()
clean_df.drop("Age", axis=1, inplace=True)
clean_df.drop("Bird Weight (g)", axis=1, inplace=True)

# removing Visible Oil Assessment
# removing UV Light Oil Assessment
# removing Oil at Capture Site
# These would not give in depth physiological evidence for whether the 
# bird was in an oil spill. The current values would likely lead to our model
# classifying all records with any value other than "None" as impacted, when
# our goal is to identify physiological markers of impacted birds, rather
# than physical markers.
clean_df.groupby(["Species", "Classification"])["Visible Oil Assessment"].describe()
clean_df.groupby(["Species", "Classification"])["UV Light Oil Assessment"].describe()
clean_df.groupby(["Species", "Classification"])["Oil at Capture Site"].describe()
clean_df.drop("Visible Oil Assessment", axis=1, inplace=True)
clean_df.drop("UV Light Oil Assessment", axis=1, inplace=True)
clean_df.drop("Oil at Capture Site", axis=1, inplace=True)

# removing RBC Count (million/cubic mm)
# Unable to resolve NaNs and combine both discrete and continous data that
# is found in the column.
clean_df.groupby(["Species", "Classification"])["RBC Count (million/cubic mm)"].describe()
clean_df.drop("RBC Count (million/cubic mm)", axis=1, inplace=True)

# removing Polychromasia
# The top frequency for all records is slight, meaning this column is unlikely
# to contain data useful to our model.
clean_df.groupby(["Species", "Classification"])["Polychromasia"].describe()
clean_df.drop("Polychromasia", axis=1, inplace=True)

# removing Anisocytosis
# The top frequency for all records is slight, meaning this column is unlikely
# to contain data useful to our model.
clean_df.groupby(["Species", "Classification"])["Anisocytosis"].describe()
clean_df.drop("Anisocytosis", axis=1, inplace=True)

# removing Serum Condition
# This column is simply a "general description" String, which cannot
# be normalized into discrete bins. It may be possible to perform
# sentiment analysis on these Strings to give our classification model,
# however this seems beyond the scope of this project.
clean_df.groupby(["Species", "Classification"])["Serum Condition"].describe()
clean_df.drop("Serum Condition", axis=1, inplace=True)

# removing Buffy Coats
# The data is sparse, therefore we cannot pull enough useful information out
# of this column. It is likely that the researchers could not obtain
# enough whole blood for proper analysis
clean_df["Buffy Coat"].isna().sum()
clean_df.drop("Buffy Coat", axis=1, inplace=True)

# removing Date
# This column doesn't provide any relevant physiological data
clean_df.drop("Date", axis=1, inplace=True)

# removing Capture Site 
# Can get the location (Capture Site data) via latitude and longitude, so 
# this String does not help us
clean_df.drop("Capture Site", axis=1, inplace=True)

# removing State 
# Can get the State data via latitude and longitude
clean_df.drop("State", axis=1, inplace=True)

# There are no more NaN values in our cleaned dataframe
clean_df.isna().sum()

# Dropping band number as it is just a key 
# and is not useful for actual prediction
clean_df.drop("Band Number", axis=1, inplace=True)

# Dropping longitude and latitude because they are not physiological markers,
# and we do not want to overfit the model to our particular geographical data
clean_df.drop("Longitude", axis=1, inplace=True)
clean_df.drop("Latitude", axis=1, inplace=True)

In [0]:
# One-hot encoding Species, as it is our only categorical variable
encoder = OneHotEncoder(sparse = False)

species = clean_df["Species"].values.reshape(-1, 1)

encoded_species = encoder.fit_transform(species)
encoded_species_names = [species[3:] for species in encoder.get_feature_names()]
encoded_species_df = pd.DataFrame(encoded_species, columns = encoded_species_names)

clean_df_with_species = clean_df.copy()
clean_df.drop("Species", axis=1, inplace=True)
clean_df = clean_df.join(encoded_species_df)

# Encoding classification as a boolean
clean_df["Classification"] = clean_df["Classification"].apply(lambda x: 0 if x == "Reference" else 1)

Now that our data is _clean_, we define our features and target below:


### 3.2. Data Exploration
* Generate appropriate data visualizations for your key variables identified in the previous section
* You should have at least three visualizations (and at least two different visualization types)
* For each visualization provide an explanation regarding the variables involved and an interpretation of the graph.
* Hypothesis testing
* If you are using Plotly, insert your visualizations as images as well (upload the graph images to an online source, e.g. github, and link those in Jupyter Notebook)


In [0]:
features = clean_df.drop("Classification", axis=1)
target = clean_df["Classification"]
features

Heinz Body Count  Reticulocytes Count  ...  CLRA  GREG 
0                 0.0                 43.0  ...   0.0    0.0
1                 0.0                 45.0  ...   0.0    0.0
2                 0.0                 42.5  ...   0.0    0.0
3                 0.0                 42.5  ...   0.0    0.0
4                 0.0                 42.5  ...   0.0    0.0
..                ...                  ...  ...   ...    ...
579               0.0                 36.0  ...   0.0    1.0
580               0.0                 31.0  ...   0.0    1.0
581               0.0                 47.0  ...   0.0    1.0
582               0.0                 41.0  ...   0.0    1.0
583               0.0                 30.0  ...   0.0    1.0

[584 rows x 34 columns]

###Visualization 1: PAH vs Classification
![Visualization 1](https://i.imgur.com/9uSGGXZ.png)

PAH stands for Polycyclic Aromatic Hydrocarbons. As a result of the information discovered in our background research, we hypothesized that PAH levels will be higher in impacted birds than reference birds. Thus, we wanted to visualize PAH values and see if this appeared correct. 


In [0]:
fig = px.scatter(clean_df, x="Classification", y="PAH (ng/ml)", color=species)
fig.update_layout(xaxis_type='category')
fig.show()

###Visualization 2: Reticulocytes Count vs Classification
![Visualization 2](https://i.imgur.com/d4QWedH.png)

Reticulocytes are immature RBCs (Red Blood Cells). As a result of the information discovered in our background research, we hypothesized that Reticulocyte counts will be lower in impacted birds than reference birds. Thus, we wanted to visualize this relationship.


In [0]:
fig = px.bar(clean_df, x="Classification", y=sorted(clean_df["Reticulocytes Count"]), color=species, barmode='group', labels={"y" : "Reticulocytes Count"})
fig.update_layout(xaxis_type='category')
fig.show()

###Visualization 3: Uric Acid vs Classification
![Visualization 3](https://i.imgur.com/NoYQqtw.png)

Uric Acid is the main excretory product in birds, reptiles, and insects. Again, our background research led us to our hypothesis that Uric Acid levels will be lower in impacted birds than reference birds.

In [0]:
fig = px.bar(clean_df, x="Classification", y=sorted(clean_df["Uric Acid (mg/dl)"]), color=species, barmode='group', labels={"y" : "Uric Acid (mg/dl)"})
fig.update_layout(xaxis_type='category')
fig.show()

### 3.3. Feature Selection & Model Construction
* Apply machine learning classifier algorithms 
* Evaluate models to determine which model to base the feature selection on
* Select features using model-based feature selection

Next, we define all the classifiers we will use (since we have a classification project), and we will use both cross validation (KFold) and `train_test_split` to split our _scaled_ features, then train the algorithms on the _scaled_ features, then evaluate metrics in those algorithms. We will also hypertune our parameters in each model to avoid overfitting/underfitting to our data.


In [0]:
classifiers = {
    "kNN" : KNeighborsClassifier(n_neighbors=3),
    "SVM" : LinearSVC(max_iter=100000, random_state=42069),
    "GaussianNB" : GaussianNB(),
    "Tree" : DecisionTreeClassifier(random_state=42069),
    "Logic" : LogisticRegression(random_state=42069)
}

In [0]:
# define MinMaxScaler
scaler = MinMaxScaler()

# scale features first, before we use multiple model selection
scaler.fit(features)
features_scaled = scaler.transform(features)

# KFold for cross validation
kfold = KFold(n_splits=10, random_state=42069, shuffle=True)

# test all above models to see which one is best for our data
for classifier_name, classifier_object in classifiers.items():
  # array of correct predictions over total predictions for each fold 
  scores = cross_val_score(estimator=classifier_object, X=features_scaled, y=target, cv=kfold)


  print(classifier_name + ": \n\t" 
        # mean of distribution of accuracies by fold
        + f'mean accuracy={scores.mean():.2%} ' 
        # standard deviation of distribution of fold accuracies
        + f'standard deviation={scores.std():.2%}' + "\n")

# split data into training and testing set
X_train, X_test, y_train, y_test = train_test_split(features, target, random_state=42069)

# scale training and testing set
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

for classifier_name, classifier_object in classifiers.items():
  
  classifier_object.fit(X=X_train_scaled, y=y_train)

  predicted = classifier_object.predict(X=X_test_scaled)

  # percent of correct predictions over total predictions
  accuracy = classifier_object.score(X_test_scaled, y_test)
  print(classifier_name + ": \n\t" + f'Predicted accuracy on the test data: {accuracy*100:.2f}%' + "\n")

kNN: 
	mean accuracy=91.09% standard deviation=1.71%

SVM: 
	mean accuracy=88.85% standard deviation=3.42%

GaussianNB: 
	mean accuracy=67.81% standard deviation=5.49%

Tree: 
	mean accuracy=88.19% standard deviation=3.07%

Logic: 
	mean accuracy=84.92% standard deviation=2.60%

kNN: 
	Predicted accuracy on the test data: 91.10%

SVM: 
	Predicted accuracy on the test data: 91.78%

GaussianNB: 
	Predicted accuracy on the test data: 66.44%

Tree: 
	Predicted accuracy on the test data: 92.47%

Logic: 
	Predicted accuracy on the test data: 84.93%



Based on these results, we will use the Decision Tree Classifier, since it
gave us the best accuracy and a relatively small and consistent standard deviation with and without hyperparameter tuning. We considered using kNN, as it too had a high accuracy and low standard deviation, although this result was only after tuning n_neighbors to 3, which increases the risk of underfitting the model to our data. To prevent this risk, we will stick to Decision Trees, as this gave us the most reliable accuracy and standard deviation, even when hyperparameter tuning max_depth. We also attempted to hypertune the other classifiers, but none gave more accurate results compared to the defaults.

Next, we will use the Decision Trees model to help us decide which features to keep. 

In [0]:
# Feature Selection

# use Decision Tree model selection
select = SelectFromModel(DecisionTreeClassifier(random_state=42069))

# fit model selection to scaled training data
select.fit(X_train_scaled, y_train)

# Get selected features in training and testing set
# based on model selection output
X_train_selected = select.transform(X_train_scaled)
X_test_selected = select.transform(X_test_scaled)

# display our selected columns
selected_columns = []
for feature, selected in zip(features.columns, select.get_support()):
  if selected:
    selected_columns.append(feature)
selected_columns

# make a dataframe with only our selected features
selected_features = clean_df[selected_columns]
selected_features

Reticulocytes Count  Packed Cell Volume (%)  ...  Glucose  Cholesterol (mg/dl)
0                   43.0                    48.5  ...    315.0                200.0
1                   45.0                    47.0  ...    358.0                235.0
2                   42.5                    45.0  ...    310.0                160.0
3                   42.5                    47.0  ...    324.0                168.0
4                   42.5                    46.0  ...    412.0                215.0
..                   ...                     ...  ...      ...                  ...
579                 36.0                    45.0  ...    260.0                176.0
580                 31.0                    46.0  ...    284.0                204.0
581                 47.0                    44.0  ...    251.0                182.0
582                 41.0                    45.0  ...    291.0                186.0
583                 30.0                    46.0  ...    257.0                205.0

[584 rows x 8 columns]

In [0]:
# split data into training and testing set
X_train, X_test, y_train, y_test = train_test_split(selected_features, target, random_state=42069)

# scale selected features based on training data
scaler = MinMaxScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)
selected_features = scaler.transform(selected_features)
selected_features

array([[0.14150943, 0.63291139, 0.0270217 , ..., 0.04918033, 0.61132812,
        0.36363636],
       [0.16037736, 0.59493671, 0.0270217 , ..., 0.09836066, 0.6953125 ,
        0.47727273],
       [0.13679245, 0.5443038 , 0.0270217 , ..., 0.14139344, 0.6015625 ,
        0.23376623],
       ...,
       [0.17924528, 0.51898734, 0.        , ..., 0.44672131, 0.48632812,
        0.30519481],
       [0.12264151, 0.5443038 , 0.        , ..., 0.91803279, 0.56445312,
        0.31818182],
       [0.01886792, 0.56962025, 0.        , ..., 0.32377049, 0.49804688,
        0.37987013]])

Now we will perform our hypothesis tests:

In [0]:
# get a list of unique species
unique_species = []
for bird in species[:-1]:
  unique_species.append(bird[0])
unique_species=list(set(unique_species))

# function to run an independent samples t-test to determine if 
# a feature is significantly different among reference and impacted 
# groups of all species
def t_test_by_species(df_, column_):
  for bird in unique_species:
    species_df = df_[df_["Species"] == bird]
    reference = species_df[species_df["Classification"] == "Reference"][column_]
    impact = species_df[species_df["Classification"] == "Impacted"][column_]
    t_test = stats.ttest_ind(reference, impact)
    deg_f = len(reference) + len(impact) - 2
    print(f'Species: \"{bird}\"')
    print(f'\tdf={deg_f}')
    print(f'\ttest statistic={t_test.statistic}')
    print(f'\tp-value={t_test.pvalue}')
    print(f'\tReference Stats:')
    print(f'\t\tN={len(reference)}')
    print(f'\t\tmean={reference.mean()}')
    print(f'\t\tstd err={reference.sem()}')
    print(f'\tImpacted Stats:')
    print(f'\t\tN={len(impact)}')
    print(f'\t\tmean={impact.mean()}')
    print(f'\t\tstd err={impact.sem()}')
    if t_test.pvalue <= 0.05:
      print(f'These results indicate that the {"Impacted" if impact.mean() > reference.mean() else "Reference"} group has a significantly higher {column_} than the {"Reference" if impact.mean() > reference.mean() else "Impacted"} group.')
    else: 
      print(f'These results indicate that there is no significant difference between the {column_} means of the Impacted and Reference groups.')
    print()

### PAH

In [0]:
t_test_by_species(clean_df_with_species, "PAH (ng/ml)")

Species: "BLSK"
	df=140
	test statistic=-1.3890870653826484
	p-value=0.16701187036617865
	Reference Stats:
		N=73
		mean=2.184931506849315
		std err=0.9781724939021627
	Impacted Stats:
		N=69
		mean=7.585217391304348
		std err=3.863260726967684
These results indicate that there is no significant difference between the PAH (ng/ml) means of the Impacted and Reference groups.

Species: "CLRA"
	df=178
	test statistic=-125.19469563813912
	p-value=1.8161264717308962e-175
	Reference Stats:
		N=80
		mean=0.55625
		std err=0.55625
	Impacted Stats:
		N=100
		mean=62.80000000000011
		std err=1.1425957100746156e-14
These results indicate that the Impacted group has a significantly higher PAH (ng/ml) than the Reference group.

Species: "BRPE"
	df=102
	test statistic=-1.5584595623846007
	p-value=0.12222179045987003
	Reference Stats:
		N=41
		mean=0.49024390243902444
		std err=0.4902439024390249
	Impacted Stats:
		N=63
		mean=3.5512698412698414
		std err=1.5495721704793197
These results indicate that

It appears that differences in PAH levels between Reference and Impacted birds varies by species. While for the GREG (Great Egret) species, it does not appear that PAH is significantly different among Reference vs Impacted birds, other species (notably, BLSK, or Black Skimmers) do seemingly present a difference in between the Reference and Impacted species.

When running the t-test, it is clear that 3 out of the 5 species do not support our hypothesis, as their PAH values are not significantly different from each other. Therefore, we fail to reject our null hypothesis, which is that the PAH levels between reference and impacted birds are not different.

### Reticulocytes Count

In [0]:
t_test_by_species(clean_df_with_species, "Reticulocytes Count")

Species: "BLSK"
	df=140
	test statistic=-5.082993434566469
	p-value=1.169152002225575e-06
	Reference Stats:
		N=73
		mean=53.89041095890411
		std err=1.3259730701979828
	Impacted Stats:
		N=69
		mean=66.28985507246377
		std err=2.08054986597033
These results indicate that the Impacted group has a significantly higher Reticulocytes Count than the Reference group.

Species: "CLRA"
	df=178
	test statistic=13.556375955544175
	p-value=3.215164488661006e-29
	Reference Stats:
		N=80
		mean=47.5625
		std err=0.6278967680007779
	Impacted Stats:
		N=100
		mean=39.13
		std err=0.23980421306987215
These results indicate that the Reference group has a significantly higher Reticulocytes Count than the Impacted group.

Species: "BRPE"
	df=102
	test statistic=-8.15463855047733
	p-value=9.430584359921821e-13
	Reference Stats:
		N=41
		mean=44.4390243902439
		std err=1.317355434975048
	Impacted Stats:
		N=63
		mean=62.58730158730159
		std err=1.5759708230621938
These results indicate that the Impacted g

This too appears to be species dependent. Our hypothesis does appear to be supported by the BLSK and GREG species, however, because the bars are so close together, this may not be a significant difference. Overall, there appears to be stronger support against our hypothesis, as the other 3 species (AMOY, BRPE, CLRA) seem to have more significant differences in their Reticulocyte levels, notably with the Reference classification having _lower_ values than the Impacted classification.

When running the t-test, it is clear that all of the species tested have p-values below an alpha=0.05. Therefore, all the species _are_ significantly different, although with t-tests, it is not possible to tell which distribution is "greater" than the other.  


### Uric Acid

In [0]:
t_test_by_species(clean_df_with_species, "Uric Acid (mg/dl)")

Species: "BLSK"
	df=140
	test statistic=-0.8846907114310391
	p-value=0.37784034309382164
	Reference Stats:
		N=73
		mean=6.987671232876712
		std err=0.5483958382907738
	Impacted Stats:
		N=69
		mean=7.721739130434774
		std err=0.6259413016135926
These results indicate that there is no significant difference between the Uric Acid (mg/dl) means of the Impacted and Reference groups.

Species: "CLRA"
	df=178
	test statistic=-6.996035340745777
	p-value=5.1449565606579493e-11
	Reference Stats:
		N=80
		mean=9.87625
		std err=0.334116271308604
	Impacted Stats:
		N=100
		mean=15.123999999999977
		std err=0.6151046884700023
These results indicate that the Impacted group has a significantly higher Uric Acid (mg/dl) than the Reference group.

Species: "BRPE"
	df=102
	test statistic=1.7768756641877708
	p-value=0.0785697499185139
	Reference Stats:
		N=41
		mean=19.92439024390244
		std err=5.511246572821494
	Impacted Stats:
		N=63
		mean=11.985714285714291
		std err=0.43953879579625427
These results

Again, 3 out of 5 of the species we looked at do not support our hypothesis on this, with the impacted classification having higher Uric acid values than the reference classification. Additionally, all three of those species have a notable difference in these values. However, looking at the GREG species, it _does_ seem to support our hypothesis with Uric acid values that are much higher in the reference classification that the impacted classification. However, it is important to note that reference GREG species seems to have an unusually high record for Uric acid at `239 mg/dl`, which is `211 mg/dl` higher than the next highest record in the dataset at `28 mg/dl`. This could mean that particular datapoint may not be reliable, which means this finding would not support our hypothesis. In addition, the BLSK species appears to have a slightly higher max value for Uric acid level in its reference dataset as oppose to its impacted dataset, although this difference cannot be said to be significant. Therefore, this graph most likely does not support our hypothesis.

When running the t-test, it is clear that 3 out of the 5 species do not support our hypothesis, as their Uric Acid values are not significantly different from each other. Therefore, we fail to reject our null hypothesis, which is that the Uric Acid levels between reference and impacted birds are not different.

### 3.4. Model Evaluation & Optimization
* Evaluate the performance of your algorithms based on scores from KFold cross-validation as well as the percentage-split approach
* Interpret your results from multiple models
* Tune your models using appropriate hyperparameters
* Explain why you are doing this (e.g., to avoid overfitting, etc.)

Now, we will rerun our model evaluation/metrics, as we have selected appropriate features. We do this to ensure that post-feature selection, the models still exhibit the same accuracy. We are especially looking at kNN, as it was a fairly accurate model the first time around. We will also be hyperparameter tuning again, just to ensure that none of the models have surpassed Decision Trees in accuracy.

In [0]:
classifiers = {
    "kNN" : KNeighborsClassifier(n_neighbors=4),
    "SVM" : LinearSVC(max_iter=100000, random_state=42069, dual=False),
    "GaussianNB" : GaussianNB(),
    "Tree" : DecisionTreeClassifier(random_state=42069),
    "Logic" : LogisticRegression(max_iter=100000, random_state=42069, dual=False)
}

In [0]:
# test all above models to see which one is best for our data
for classifier_name, classifier_object in classifiers.items():
  kfold = KFold(n_splits=10, random_state=42069, shuffle=True)

  scores = cross_val_score(estimator=classifier_object, X=selected_features, y=target, cv=kfold)

  print(classifier_name + ": \n\t" + f'mean accuracy={scores.mean():.2%} ' + f'standard deviation={scores.std():.2%}' + "\n")

for classifier_name, classifier_object in classifiers.items():
  
  classifier_object.fit(X=X_train_scaled, y=y_train)

  predicted = classifier_object.predict(X=X_test_scaled)

  accuracy = classifier_object.score(X_test_scaled, y_test)
  print(classifier_name + ": \n\t" + f'Predicted accuracy on the test data: {accuracy*100:.2f}' + "\n")

kNN: 
	mean accuracy=90.58% standard deviation=3.09%

SVM: 
	mean accuracy=82.38% standard deviation=4.92%

GaussianNB: 
	mean accuracy=71.41% standard deviation=5.18%

Tree: 
	mean accuracy=90.24% standard deviation=4.98%

Logic: 
	mean accuracy=72.98% standard deviation=5.05%

kNN: 
	Predicted accuracy on the test data: 91.10

SVM: 
	Predicted accuracy on the test data: 80.82

GaussianNB: 
	Predicted accuracy on the test data: 71.23

Tree: 
	Predicted accuracy on the test data: 91.78

Logic: 
	Predicted accuracy on the test data: 67.81



After rerunning our metrics post-feature selection we discovered that Decision Trees are still our best model, as it had the highest score when using cross validation, and it had the highest predicted accuracy in the class report when using `train_test_split`. We attempted to hypertune each algorithm's parameters once again, and discovered that this time around, kNN appeared more accurate with `n_neighbors=4`, rather than 3. However, it still was not as accurate as Decision Trees.

### 3.5. Model Testing
* Test your tuned algorithms using your testing set
* Evaluate performance using additional evaluation metrics such as the confusion matrix and the classification report.

In [0]:
# Get our most accurate classifier
tree = classifiers.get("Tree")
predicted = tree.predict(X=X_test_scaled)

# Get and print the confusion matrix
confusion = confusion_matrix(y_true = y_test, y_pred = predicted)
confusion_df = pd.DataFrame(confusion, 
                            index=["Actual reference", "Actual impacted"], 
                            columns=["Predicted reference", "Predicted impacted"])
confusion_df

Predicted reference  Predicted impacted
Actual reference                   60                   5
Actual impacted                     7                  74

In [0]:
# get our classification report
class_report = classification_report(y_true = y_test, y_pred = predicted)
print(class_report)

              precision    recall  f1-score   support

           0       0.90      0.92      0.91        65
           1       0.94      0.91      0.92        81

    accuracy                           0.92       146
   macro avg       0.92      0.92      0.92       146
weighted avg       0.92      0.92      0.92       146



Based on the confusion matrix, we can say that out of 146 records in the test set, our Decision Tree model predicted 60 reference (negative) records correctly, and 74 impacted (positive) records correctly. This means our false positive rate is `5/(74+5) = ~6.33%`. Additionally, our false negative rate is `7/(60+7) = ~10.45%`.

Based on the class report, our model has `90%` precision for 0 (i.e. ability to not label impacted records as reference records), and a `94%` precision for 1 (ability to not label reference records as impacted records). Our model has a recall of `92%` for impacted and a recall of `91%` for reference, meaning it has a `92%` chance of correctly identifying positive records and a `91%` chance of correctly identifying negative records. 

<a id="4"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 4. DISCUSSION


In Section 3.3, we ran independent sample t-tests was conducted to compare the differences in PAH (ng/mL), Reticulocyte Count, and Uric Acid (mg/dl) between the impacted and reference groups. These t-tests were run separately for each species. 

Here, the null hypothesis is that there is no difference between the means of the markers for the impacted and reference groups. The alternative hypothesis is that the markers' means are different between the impacted and reference groups.

A p-value calculated from the t-test less than and equal to 0.05 would signify that there is a significant difference between the two means, and that we will reject our null hypothesis. If the p-value is greater than 0.05, then we fail to reject our null hypothesis because the means are insignificantly different.

As stated above in Section 2, this is a supervised learning classification problem, and we approached data analysis with a hypothesis that higher values of certain physiological markers that indicate contact with oil (PAH (nh/ml), Heinz Body Counts, Reticulocytes Counts, etc.) would correspond with a higher likelihood that the given bird was impacted by an oil spill. Our data analysis also included a lot of manual analysis during the data wrangling and data exploration steps in section 3 when we were determining which features were reasonable for us to keep from the original dataset and which were not.

  After finalizing feature selection in section 3.3, we moved on to evaluate the 5 models described in section 2.3:


*   K-Nearest Neighbors-based classification (KNeighborsClassifier)
*   Linear Support Vector Classification (LinearSVC)
*   Gaussian Naive Bayes (GaussianNB)
*   Decision Tree classifier (DecisionTreeClassifier)
*   Logistic Regression classifier (LogisticRegression)


  We used the exact same method for comparing model performance in section 3.4 as we did in section 3.3: using the aforementioned models, we iteratively tested each one using both cross validation with KFold and a standard 75/25 train-test split. The metric used to evaluate the models was the prediction accuracy on the test data for train-test split, and the average of prediction accuracies for the cross validation. We also performed hyperparameter tuning in this section to determine whether or not any of the models could benefit from it, and we did discover that the K-Nearest Neighbors classifier benefited most from a value of n_neighbors=4, however even after this change the Decision Tree classifier was still our most accurate model.

  Once we selected the Decision Tree classifier as the final candidate for our predictive model we generated a confusion matrix and a classification report using the model’s predictions. The results of the classification report and the confusion matrix indicated that our model had a final accuracy rate of roughly 92% on the testing data, with precision/recall scores for positive and negative classifications being 94%/91% for positive and 90%/92% for negative respectively.

  In conclusion, our attempts to train a classifier on the selected dataset to identify whether a given bird was affected by an oil spill were successful, and we were able to use our model to accurately predict whether a bird was classified as ‘impacted’ or ‘reference’. We accomplished this after a great deal of data wrangling and feature engineering performed on our dataset to ensure that the data was clean enough for our model to use it. For future work concerning this topic, we would like to look towards working with a larger, more complete set of data. Although the original dataset contained over 1200 rows, we ended up removing more than half of them due to the pervasiveness of ‘NaN’ and ‘None’ values in the dataset, with certain species of birds having particularly sparse data. Additionally, some feature variables that were potentially useful for training the model—such as Bird Weight (g)—had a significant number of values that were seemingly nonsensical/inaccurate and ultimately made it impossible to use those variables as features for the model. Having complete and correct values for these columns in the original dataset would be another improvement that could allow future work on the same dataset to see if the model could be improved upon.


<a id="5"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

### CONTRIBUTIONS
The introduction (Section 1) was written by Tom Cairns. The methodology (Section 2) was written by Amy Tang. Our work to clean up the data, run hypothesis tests, scale our data, select features, train machine learning models, and test our machine learning model (Section 3.1, 3.3-3.5) was done as a group in a collaborative-programming manner. The visualizations (Section 3.2) were generated by Ben Kosiborod. The discussion section (Section 4) was written by Gabe Deacon. The report was edited by everyone as a collaborative group.
